src/main/java/org/opensearch/knn/profiler/SegmentProfilerState.java

```
@@ -0,0 +1,141 @@
+        KNNVectorValues<?> vectorValues = knnVectorValuesSupplier.get();
+
+        if (vectorValues == null) {
+            log.info("No vector values available");
+            return new SegmentProfilerState(new ArrayList<>());
+        }
+
+        // Initialize vector values
+        KNNCodecUtil.initializeVectorValues(vectorValues);
+        List<SummaryStatistics> statistics = new ArrayList<>();
+
+        // Return empty state if no documents are present
+        if (vectorValues.docId() == NO_MORE_DOCS) {
+            log.info("No vectors to profile");
+            return new SegmentProfilerState(statistics);
+        }
+
+        int dimension = vectorValues.dimension();
+        log.info("Starting vector profiling with dimension: {}", dimension);
+
+        // Initialize statistics collectors for each dimension
+        for (int i = 0; i < dimension; i++) {
+            statistics.add(new SummaryStatistics());
+        }
+
+        // Process all vectors
+        int vectorCount = 0;
+        for (int doc = vectorValues.docId(); doc != NO_MORE_DOCS; doc = vectorValues.nextDoc()) {
+            vectorCount++;
+            Object vector = vectorValues.getVector();
+            processVectors(vector, statistics);
+        }
+
+        log.info("Vector profiling completed - processed {} vectors", vectorCount);
+        logDimensionStatistics(statistics, dimension);
+
+        return new SegmentProfilerState(statistics);
+    }
+
+    /**
+     * Helper method to process a vector and update statistics
+     * @param vector
+     * @param statistics
+     */
+    private static void processVectors(Object vector, List<SummaryStatistics> statistics) {
+        if (vector instanceof float[]) {
+            processFloatVector((float[]) vector, statistics);
+        } else if (vector instanceof byte[]) {
+            processByteVector((byte[]) vector, statistics);
+        } else {
+            throw new IllegalArgumentException("Unsupported vector type.");
```

Should we throw an exception here if we aren't able to profile a vector of a unsupported type? This is a profiler that wouldn't be in the hot path - would it be better to log and move on?

---

Understood, creating log to catch issues